File: leetcode/0059-spiralMatrixII.js
/*
https://leetcode.com/problems/spiral-matrix-ii/

Given a positive integer n, generate an n x n matrix filled with elements from 1 to n^2 in spiral order.

Input: n = 3
Output: [[1,2,3],[8,9,4],[7,6,5]]
*/

// Runtime: 85 ms	Memory: 42.3 MB
const generateMatrix = (n) => {
  let num = 1;
  let rowStart = 0;
  let colStart = 0;
  let rowEnd = n - 1;
  let colEnd = n - 1;
  const result = [];

  for (let i = 0; i < n; i++) {
    result.push([]);
  }

  while (rowStart <= rowEnd && colStart <= colEnd) {
    // North
    for (let i = colStart; i <= colEnd; i++) {
      result[rowStart][i] = num;
      num++;
    }

    // East
    rowStart++;
    for (let i = rowStart; i <= rowEnd; i++) {
      result[i][colEnd] = num;
      num++;
    }

    // South
    colEnd--;
    for (let i = colEnd; i >= colStart; i--) {
      result[rowEnd][i] = num;
      num++;
    }

    // West
    rowEnd--;
    for (let i = rowEnd; i >= rowStart; i--) {
      result[i][colStart] = num;
      num++;
    }
    colStart++;
  }

  return result;
};

console.log(generateMatrix(3));

File: leetcode/0059-spiralMatrixII.py
from typing import List

# Runtime: 73 ms	Memory: 13.9 MB
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        idx = 1
        row_start = 0
        col_start = 0
        row_end = n - 1
        col_end = n - 1
        result = [[None] * n for _ in range(n)]
        print(result)

        while row_start <= row_end and col_start <= col_end:
            # North
            for col in range(col_start, col_end + 1):
                result[row_start][col] = idx
                idx += 1

            # East
            row_start += 1
            for row in range(row_start, row_end + 1):
                result[row][col_end] = idx
                idx += 1

            # South
            col_end -= 1
            for col in range(col_end, col_start - 1, -1):
                result[row_end][col] = idx
                idx += 1

            # West
            row_end -= 1
            for row in range(row_end, row_start - 1, -1):
                result[row][col_start] = idx
                idx += 1
            col_start += 1

        return result


print(Solution().generateMatrix(3))
